<a href="https://colab.research.google.com/github/Candice-Chuang/APRD-6342-Digital-Ad/blob/master/Python_Exercise_1_Basic_Ad_Analyses_Candice.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [0]:
import pandas as pd

DATA_URL = 'http://128.138.93.164/aprd6342/data/TravelPonyFacebook.csv'
df = pd.read_csv(DATA_URL)
df



Start Date  ... Cost per Page Engagement (USD)
0     2013-12-02  ...                       1.219512
1     2013-12-01  ...                       1.395349
2     2013-11-30  ...                       1.190476
3     2013-11-29  ...                       1.167315
4     2013-11-22  ...                       2.512857
...          ...  ...                            ...
3700  2013-06-11  ...                       1.050000
3701  2013-06-10  ...                       0.533333
3702  2013-06-09  ...                       0.990000
3703  2013-06-08  ...                       0.316667
3704  2013-06-07  ...                       0.000000

[3705 rows x 20 columns]

In [0]:
pip install -U arrow

     |████████████████████████████████| 51kB 3.9MB/s 


In [0]:
df['Start Date'] = pd.to_datetime(df['Start Date'])
df['day_of_week'] = df['Start Date'].dt.day_name()
df

Start Date    End Date  ... Cost per Page Engagement (USD) day_of_week
0    2013-12-02  2013-12-02  ...                       1.219512      Monday
1    2013-12-01  2013-12-01  ...                       1.395349      Sunday
2    2013-11-30  2013-11-30  ...                       1.190476    Saturday
3    2013-11-29  2013-11-29  ...                       1.167315      Friday
4    2013-11-22  2013-11-22  ...                       2.512857      Friday
...         ...         ...  ...                            ...         ...
3700 2013-06-11  2013-06-11  ...                       1.050000     Tuesday
3701 2013-06-10  2013-06-10  ...                       0.533333      Monday
3702 2013-06-09  2013-06-09  ...                       0.990000      Sunday
3703 2013-06-08  2013-06-08  ...                       0.316667    Saturday
3704 2013-06-07  2013-06-07  ...                       0.000000      Friday

[3705 rows x 21 columns]

In [0]:
df['cost_per_impression'] = df['Amount Spent (USD)']/df['Impressions']
num_day = df.groupby('day_of_week').count()
n_d = num_day.loc[:, 'Start Date']





In [0]:
sum_day = df.groupby('day_of_week').sum()
s_d = sum_day.loc[:, 'cost_per_impression']


In [0]:
ndf = pd.concat([n_d,s_d],axis=1,join_axes=[n_d.index])
ndf

ndf['avg_impression'] = ndf['cost_per_impression']/ndf['Start Date']   #sum all cost per impression/ numbers of days
ndf

ndf.sort_values(by=['avg_impression'],ascending=False)

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:1: FutureWarning: The join_axes-keyword is deprecated. Use .reindex or .reindex_like on the result to achieve the same functionality.
  """Entry point for launching an IPython kernel.


Start Date  cost_per_impression  avg_impression
day_of_week                                                 
Friday              502             2.056639        0.004097
Sunday              476             1.755447        0.003688
Thursday            574             1.950631        0.003398
Tuesday             474             1.440979        0.003040
Wednesday           591             1.761665        0.002981
Monday              589             1.713976        0.002910
Saturday            499             1.311720        0.002629

CONCLUSION:

Frm the analytics result, we knows that "Firday" has the highest cost per impression and "Saturday" has the lowest cost per impression. That is to say,it cost most on Friday to attract customers and on Saturday it cost less.


In [0]:
relation = df[['Amount Spent (USD)','Reach','Frequency','Unique Clicks','Page Likes']]
relation



Amount Spent (USD)   Reach  Frequency  Unique Clicks  Page Likes
0                 300.00  278117   2.314303            313           3
1                 300.00  221255   2.079090            348           6
2                 300.00  267915   2.499069            345           2
3                 300.00  280332   2.365406            275           3
4                  35.18   14269   1.000420             61           2
...                  ...     ...        ...            ...         ...
3700                1.05     298   1.003356              1           0
3701                1.60     514   1.060311              3           0
3702                0.99     111   1.000000              1           0
3703                0.95     143   1.000000              3           0
3704                2.25    4711   1.087243              1           0

[3705 rows x 5 columns]

In [0]:
relation.cov()  

Amount Spent (USD)  ...   Page Likes
Amount Spent (USD)          666.382402  ...    51.294872
Reach                    173984.356384  ...  7652.619815
Frequency                     0.416286  ...     0.000059
Unique Clicks               591.888877  ...    39.815824
Page Likes                   51.294872  ...     6.879087

[5 rows x 5 columns]

In [0]:
relation.corr()

Amount Spent (USD)     Reach  ...  Unique Clicks  Page Likes
Amount Spent (USD)            1.000000  0.703124  ...       0.882993    0.757612
Reach                         0.703124  1.000000  ...       0.722249    0.304388
Frequency                     0.130201  0.334101  ...       0.135103    0.000182
Unique Clicks                 0.882993  0.722249  ...       1.000000    0.584614
Page Likes                    0.757612  0.304388  ...       0.584614    1.000000

[5 rows x 5 columns]

Conclusion

From the correlation matrix above, "Unique Clicks" has the highest correlation with Amount Spent, which means "Unique Clicks" has strong linear correlation with the regression model(with independent variable Amount Spent). 

In [0]:
import statsmodels.api as sm
import numpy as np
import pandas as pd


import numpy
from pandas import read_csv
from matplotlib import pyplot as plt
from sklearn.linear_model import LinearRegression


data = df
data


Start Date    End Date  ... day_of_week cost_per_impression
0    2013-12-02  2013-12-02  ...      Monday            0.000466
1    2013-12-01  2013-12-01  ...      Sunday            0.000652
2    2013-11-30  2013-11-30  ...    Saturday            0.000448
3    2013-11-29  2013-11-29  ...      Friday            0.000452
4    2013-11-22  2013-11-22  ...      Friday            0.002464
...         ...         ...  ...         ...                 ...
3700 2013-06-11  2013-06-11  ...     Tuesday            0.003512
3701 2013-06-10  2013-06-10  ...      Monday            0.002936
3702 2013-06-09  2013-06-09  ...      Sunday            0.008919
3703 2013-06-08  2013-06-08  ...    Saturday            0.006643
3704 2013-06-07  2013-06-07  ...      Friday            0.000439

[3705 rows x 22 columns]

In [0]:
lrModel = LinearRegression()

x = data[['Frequency','Reach']]
y = data[['Unique Clicks']]

lrModel.fit(x,y)

model = sm.OLS(y, x).fit()
predictions = model.predict(x)

model.summary()



<class 'statsmodels.iolib.summary.Summary'>
"""
                                 OLS Regression Results                                
=======================================================================================
Dep. Variable:          Unique Clicks   R-squared (uncentered):                   0.557
Model:                            OLS   Adj. R-squared (uncentered):              0.556
Method:                 Least Squares   F-statistic:                              2325.
Date:                Wed, 13 Nov 2019   Prob (F-statistic):                        0.00
Time:                        23:17:29   Log-Likelihood:                         -15973.
No. Observations:                3705   AIC:                                  3.195e+04
Df Residuals:                    3703   BIC:                                  3.196e+04
Df Model:                           2                                                  
Covariance Type:            nonrobust                                                  
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Frequency      3.6139      0.298     12.109      0.000       3.029       4.199
Reach          0.0019   3.12e-05     62.490      0.000       0.002       0.002
==============================================================================
Omnibus:                     5107.616   Durbin-Watson:                   0.803
Prob(Omnibus):                  0.000   Jarque-Bera (JB):          5128392.747
Skew:                           7.331   Prob(JB):                         0.00
Kurtosis:                     184.674   Cond. No.                     9.84e+03
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 9.84e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

Conclusion



From the Regression model result, we knows that R square(0<R squared<1) is 0.557, means the model is good at predict the relationship between dependent variable(Unique Clicks) and independent variables(Frequency and Reach). In addition, for both independent variables, both p-value are less than 0.5 that means both variables have significant effect on the model. Variable "Frequency" has the highest coefficient(3.61319), which means Frequency has strong linear relationship with the model.Thus,variable "Frequency" most strongly predicts Unique clicks
